pkg/gadgets/snapshot/socket/tracer/tracer_test.go

```diff
@@ -0,0 +1,151 @@
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	tracer, err := NewTracer(types.ALL)
+	require.ErrorIsf(t, nil, err, "creating tracer: %v", err)
+
+	tracer.CloseIters()
+}
+
+type testCase struct {
+	name          string
+	proto         types.Proto
+	addr          string
+	port          int
+	expectedEvent types.Event
+	socketCreator func(addr string, port int) error
+}
+
+func TestSnapshotSocket(t *testing.T) {
+	t.Parallel()
+
+	utilstest.RequireRoot(t)
+
+	cases := []testCase{
+		{
+			name:  "tcp_v4",
+			proto: types.TCP,
+			addr:  "127.0.0.1",
+			port:  8082,
+			expectedEvent: types.Event{
+				Protocol: "TCP",
+				Status:   "LISTEN",
+			},
+			socketCreator: func(addr string, port int) error {
+				conn, err := net.Listen("tcp", fmt.Sprintf("%s:%d", addr, port))
+				if err != nil {
+					return fmt.Errorf("listening to %s: %w", addr, err)
+				}
+				t.Cleanup(func() { conn.Close() })
+
+				return nil
+			},
+		},
+		{
+			name:  "udp_v4",
+			proto: types.UDP,
+			addr:  "127.0.0.1",
+			port:  8082,
+			expectedEvent: types.Event{
+				Protocol: "UDP",
+				Status:   "INACTIVE",
+			},
+			socketCreator: func(addr string, port int) error {
+				conn, err := net.ListenUDP("udp", &net.UDPAddr{
+					Port: port,
+					IP:   net.ParseIP(addr),
+				})
+				if err != nil {
+					return fmt.Errorf("listening to %s: %w", addr, err)
+				}
+				t.Cleanup(func() { conn.Close() })
+
+				return nil
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			t.Parallel()
+
+			runner := utilstest.NewRunnerWithTest(t, nil)
+			utilstest.RunWithRunner(t, runner, func() error {
+				return c.socketCreator(c.addr, c.port)
+			})
+
+			tracer, err := NewTracer(c.proto)
+			require.ErrorIsf(t, nil, err, "creating tracer: %v", err)
+			defer tracer.CloseIters()
+
+			evs, err := tracer.RunCollector(uint32(runner.Info.Tid), "", "", "")
+			require.ErrorIsf(t, nil, err, "running collector: %v", err)
+
+			type extra struct {
+				addr string
+				port uint16
+			}
+
+			events := make([]types.Event, len(evs))
+			for i, ev := range evs {
+				events[i] = *ev
+
+				// This is hard to guess the inode number, let's normalize it for the
+				// moment.
+				events[i].InodeNumber = 0
+			}
+
+			utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, e extra) *types.Event {
+				// Deal with common fields here.
+				c.expectedEvent.Event = eventtypes.Event{Type: eventtypes.NORMAL}
+				c.expectedEvent.SrcEndpoint = eventtypes.L4Endpoint{
+					L3Endpoint: eventtypes.L3Endpoint{
+						Addr: e.addr,
+					},
+					Port: e.port,
+				}
+				c.expectedEvent.DstEndpoint = eventtypes.L4Endpoint{
+					L3Endpoint: eventtypes.L3Endpoint{
+						// There is no connection in this test, so there remote address is null.
+						Addr: "0.0.0.0",
+					},
+				}
+				c.expectedEvent.NetNsID = runner.Info.NetworkNsID
+
+				return &c.expectedEvent
+			})(t, nil, extra{c.addr, uint16(c.port)}, events)
```

Comment:
I think this function should be in the test definition (like [here](https://github.com/inspektor-gadget/inspektor-gadget/blob/70010a391b9d5b9f908bc75d5da08b8c03ef7ceb/pkg/gadgets/internal/socketenricher/tracer_test.go#L77)). I agree this will cause some more code duplication but I think the extra flexibility is worth it.